ir/src/main/scala/org/scalajs/ir/Trees.scala

```diff
@@ -1059,29 +1062,131 @@ object Trees {
+      val ordinal: Int) // intentionally public
+      extends AnyVal {
+
+    import MemberNamespace._
+
+    def isStatic: Boolean = (ordinal & StaticFlag) != 0
+
+    def isPrivate: Boolean = (ordinal & PrivateFlag) != 0
+
+    def isConstructor: Boolean = (ordinal & ConstructorFlag) != 0
+
+    def prefixString: String = this match {
+      case Public            => ""
+      case Private           => "private "
+      case PublicStatic      => "static "
+      case PrivateStatic     => "private static "
+      case Constructor       => "constructor "
+      case StaticConstructor => "static constructor "
+    }
+  }
+
+  object MemberNamespace {
+    private final val StaticFlag = 1
+
+    private final val PrivateFlag = 2
+
+    private final val ConstructorFlag = 4
+
+    final val Public: MemberNamespace =
+      new MemberNamespace(0)
+
+    final val PublicStatic: MemberNamespace =
+      new MemberNamespace(StaticFlag)
+
+    final val Private: MemberNamespace =
+      new MemberNamespace(PrivateFlag)
+
+    final val PrivateStatic: MemberNamespace =
+      new MemberNamespace(PrivateFlag | StaticFlag)
+
+    final val Constructor: MemberNamespace =
+      new MemberNamespace(ConstructorFlag)
+
+    final val StaticConstructor: MemberNamespace =
+      new MemberNamespace(ConstructorFlag | StaticFlag)
+
+    final val Count = 6
+
+    def fromOrdinal(ordinal: Int): MemberNamespace = {
+      require(0 <= ordinal && ordinal < Count,
+          s"Invalid namespace ordinal $ordinal")
+      new MemberNamespace(ordinal)
+    }
+
+    private[Trees] def fromOrdinalUnchecked(ordinal: Int): MemberNamespace =
+      new MemberNamespace(ordinal)
+
+    def forNonStaticCall(flags: ApplyFlags): MemberNamespace = {
+        if (flags.isPrivate) Private
+        else if (flags.isConstructor) Constructor
+        else Public
+      }
+
+    def forStaticCall(flags: ApplyFlags): MemberNamespace =
+      if (flags.isPrivate) PrivateStatic else PublicStatic
+  }
+
   final class MemberFlags private (val __private_bits: Int) extends AnyVal {
     import MemberFlags._
 
     @inline private def bits: Int = __private_bits
 
-    def isStatic: Boolean = (bits & StaticBit) != 0
+    def namespace: MemberNamespace =
+      MemberNamespace.fromOrdinalUnchecked(bits & NamespaceMask)
 
     def isMutable: Boolean = (bits & MutableBit) != 0
 
-    def withStatic(value: Boolean): MemberFlags =
-      if (value) new MemberFlags(bits | StaticBit)
-      else new MemberFlags(bits & ~StaticBit)
+    def withNamespace(namespace: MemberNamespace): MemberFlags =
+      new MemberFlags((bits & ~NamespaceMask) | namespace.ordinal)
 
     def withMutable(value: Boolean): MemberFlags =
       if (value) new MemberFlags(bits | MutableBit)
       else new MemberFlags(bits & ~MutableBit)
   }
 
   object MemberFlags {
-    private final val StaticShift = 0
-    private final val StaticBit = 1 << StaticShift
+    private final val NamespaceMask = 7
```

Comment:
Express this with shift as well?

Reply:
No, for this one it is important that it stays at no-shift. I have added a comment.